### ISSUE TITLE
Comparing 'likeliness' of some input against trained model

### ISSUE THREAD

**Issue by ChrisCummins** [jcjohnson/torch-rnn#68]
Hi Justin,

Apologies that I'm not much of a machine learning guy, but is it possible to use torch-rnn to compare the *likeliness* of an input text against a model trained using some corpus? So for example, with an rnn trained using Shakespeare, this function would return a low numerical score for the input 'The cat sat on the mat" (intuitively: this sentence was not likely written by Shakespeare), and a high numerical score for the input 'O Romeo, where art thou?' (intuitively: this sentence was likely written by Shakespeare). Obviously both inputs are too short to make a strong distinction, but I hope the question makes sense.

Thank you for your amazing work,
Chris

**Comment by jcjohnson** [comment]
This is not currently implemented, but it wouldn't be hard to do. During training the RNN learns a language model, which is able to assign a probability to an arbitrary sequence of tokens; by sampling from this language model we can generate new text, but if you have an existing piece of text that you want to score then you can just compute its probability under the language model.

All you need to do is run a forward pass of the trained model on your new piece of text; the probability of that text is just the training loss. To implement this, you'd need to load the model and set it to `evaluate` mode as in `sample.lua`; you'll then need to construct a `CrossEntropyCriterion` and use it to compute training loss as in `train.lua`.

**Comment by ChrisCummins** [comment]
Thanks for the prompt and complete answer! I will definitely have a go at implementing this. If successful, would you be open to a pull request with a `score.lua` script (or some more appropriate name)?

**Comment by aliabbasjp** [comment]
@ChrisCummins It is already implemented here:
https://github.com/billzorn/torch-rnn/commit/69d91a3b6d825c5836bcadf2c3d919f33ec4907c

**Comment by ChrisCummins** [comment]
Thanks for the pointer @aliabbasjp! Reading through the diff it seems like this fork has departed quite a bit from upstream. I'm assuming the usage would be something like this?

```
th train.lua -input_h5 <corpus-to-compare-likeness-against> -init_from <checkpoint-trained-on-dataset> -unk 1
```

Perhaps @billzorn could weigh-in.

**Comment by ianni67** [comment]
@aliabbasjp sorry for being dumb, but I cannot find, in the code you pointed out, the function or the options to do what @ChrisCummins asked for. Would you please, give us some more information?
Thank you very much in advance
